### DISCUSSION TITLE
DeployerPHP: use composer or docker?

### DISCUSSION THREAD

**Discussion by campaignupgrade** [deployphp/deployer#2792] [Help needed]
What are the advantages and disadvantages of using Deployer via Composer vs a Docker based system? 

I'm new to DeployerPHP and looking to use it for test and deploy stages of custom WordPress devops. It looks like I can install Deployer (and all the deps I need like WP, Bedrock, and Slack integration) or I can do the same via Docker images (we use Lando to manage containers).

I'm hoping to gather some of the community's wisdom before I put a ton of energy down the wrong path. =)

**Comment by antonmedv** [comment]
Well, Deployer is for deployments) What do you use for it?

**Reply by campaignupgrade** [reply]
I'm currently using GitHub Actions and custom bash scripts for deployment, but it's not robust enough for our projects. 

We'd like to replace a lot of Githubs actions and all of the bash script with Deployer PHP. 

Some sites recommend adding Deployer using composer. Others recommend attaching Deployer to GitHub actions using containers. I want to get community input on the two approaches.

**Reply by campaignupgrade** [reply]
Also, I hope you don't mind me mixing issues in one discussion, but is this forum the best place to find the Deployer community? Or is there a web forum or Slack channel folks use?

**Reply by antonmedv** [reply]
GitHub Discussions is our forum) Use https://github.com/deployphp/action

**Reply by campaignupgrade** [reply]
Thanks Anton. 

If I use the DeployPHP Action, then Deployer is not present on our local development stack. This seems to introduce some problems for us.

For one, Deployer for WordPress seems to include functions for pushing and pulling databases from remotes to local and vice versa; we think we want to use Deployer for this instead of our current home-grown custom bash script. Using DeployPHP Action makes this feature unavailable locally. 

Two, if Deployer is installed locally, then we can run deployments from local. If I understand Deployer correctly, it can build and test locally and push the build artifact from local to the remote. This makes it much easier to experiment and troubleshoot test and deploy systems. I've found developing complex GitHub Actions deployments to be challenging because you must trigger a new build and then review the logs to evaluate every minor change in the script. By first building functioning deployments locally, we could (in theory) then use a very simple GitHub Action that checks out and runs our deployer script.

So if I use deployer locally and not as a Github Action, I can add it to my project using `composer require deployer/dist --dev` or attach it as an image in lando (docker) using `https://hub.docker.com/r/debreczeniandras/deployerphp` or similar. 

Is this making sense or am I misunderstanding the use of the tool?

**Reply by antonmedv** [reply]
Just add Deployer as dep in composer. Or just commit deployer.phar into repo. Deployphp/action will use it.
